### PR TITLE
Fixes deletion of stopped VMs

### LINKED ISSUE #721: Can not delete a cluster with stopped VMs
Related to https://github.com/Azure/ARO-RP/issues/687

There are several flavours of this issue. We currently can not recover from cases 1 & 2.

1. When a cluster in failed state from where we can not run admin update. For example `provisioningState=Failed` and `failedProvisioningState=Deleting` ([ref](https://github.com/Azure/ARO-RP/blob/64a15e5b607bb6d77f666ab88a088994262e8197/pkg/frontend/openshiftcluster_putorpatch.go#L96-L108)).
1.  When a cluster is in succeded state, but VMs are stopped + we *can not* start them (in a suspended subscription, for example)
1.  When a cluster is in succeded state, but VMs are stopped + we *can* start them (cluster was in suspended subscription, but it's no longer a case) - should be possible to recover and delete once we have a proper fix for https://github.com/Azure/ARO-RP/issues/687

RP panics with `runtime error: index out of range [0] with length 0` here on line 30 (see full panic at the end).

https://github.com/Azure/ARO-RP/blob/64a15e5b607bb6d77f666ab88a088994262e8197/pkg/backend/openshiftcluster/delete.go#L28-L31

This happens because worker profile is a dynamic data which we receive from cluster's API server. When we can not reach the API server worker profile is not populated (`nil`).

---

```
pkg/util/recover/recover.go:19 recover.Panic() goroutine 314 [running]:
runtime/debug.Stack(0xc00042d110, 0x2, 0xc0006d1a08)
	/usr/local/go/src/runtime/debug/stack.go:24 +0x9d
github.com/Azure/ARO-RP/pkg/util/recover.Panic(0xc00042d110)
	/Users/mradchuk/workspace/src/github.com/Azure/ARO-RP/pkg/util/recover/recover.go:19 +0x9a
panic(0x65b1d80, 0xc000c73480)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/Azure/ARO-RP/pkg/backend/openshiftcluster.(*Manager).Delete(0xc002f7e300, 0x6d29f20, 0xc000c6e800, 0x1, 0x1)
	/Users/mradchuk/workspace/src/github.com/Azure/ARO-RP/pkg/backend/openshiftcluster/delete.go:30 +0xd20
github.com/Azure/ARO-RP/pkg/backend.(*openShiftClusterBackend).handle(0xc000118288, 0x6d29f60, 0xc0000d6000, 0xc00042d110, 0xc000400300, 0x0, 0x0)
	/Users/mradchuk/workspace/src/github.com/Azure/ARO-RP/pkg/backend/openshiftcluster.go:126 +0x33d
github.com/Azure/ARO-RP/pkg/backend.(*openShiftClusterBackend).try.func1(0xc00042d110, 0xc000118288, 0xc000400300)
	/Users/mradchuk/workspace/src/github.com/Azure/ARO-RP/pkg/backend/openshiftcluster.go:62 +0x121
created by github.com/Azure/ARO-RP/pkg/backend.(*openShiftClusterBackend).try
	/Users/mradchuk/workspace/src/github.com/Azure/ARO-RP/pkg/backend/openshiftcluster.go:49 +0x309
```

### PR DESCRIPTION
**Pull request by m1kola**
Fixes #721

### CODE REVIEW

File: pkg/util/azureclient/mgmt/network/securitygroups.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	mgmtnetwork "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-07-01/network"
+	"github.com/Azure/go-autorest/autorest"
+)
+
+// SecurityGroupsClient is a minimal interface for azure SecurityGroupsClient
+type SecurityGroupsClient interface {
```
Comment (jim-minter) [inline pkg/util/azureclient/mgmt/network/securitygroups.go:12]: missing the mock generation

File: pkg/backend/openshiftcluster/delete.go [hunk 2]
```diff
@@ -24,41 +23,48 @@ func (m *Manager) Delete(ctx context.Context) error {
...
-		if err != nil {
-			return err
-		}
+		for _, subnet := range *nsg.SecurityGroupPropertiesFormat.Subnets {
+			// Note: subnet only has value in the ID field,
+			// so we have to make another API request to get full subnet struct
+			// TODO: there is probably an undesirable race condition here - check if etags can help.
+			s, err := m.subnet.Get(ctx, *subnet.ID)
+			if err != nil {
+				// It is possible that we never properly had access to the subnet,
```
Comment (jim-minter) [inline pkg/backend/openshiftcluster/delete.go:45]: I think we can return an end-user facing error here now.

File: pkg/backend/openshiftcluster/delete.go [hunk 3]
```diff
@@ -24,41 +23,48 @@ func (m *Manager) Delete(ctx context.Context) error {
...
+			s.SubnetPropertiesFormat.NetworkSecurityGroup = nil
 
-		m.log.Printf("removing network security group from subnet %s", subnetID)
-		err = m.subnet.CreateOrUpdate(ctx, subnetID, s)
-		if err != nil {
-			return err
+			m.log.Printf("removing network security group from subnet %s", *s.ID)
+			err = m.subnet.CreateOrUpdate(ctx, *s.ID, s)
+			if err != nil {
+				return err
```
Comment (jim-minter) [inline pkg/backend/openshiftcluster/delete.go:66]: I think we can return an end-user facing error here now.

Comment (jim-minter) [reply pkg/backend/openshiftcluster/delete.go:66]: Probably need to return an error of the form "failed to update subnet ... nested error was ..."

**Comment by m1kola** [comment]
All tests are green - GitHub just didn't receive status (because of [this incident](https://www.githubstatus.com/incidents/6tcfpztf6j9m), I think)

File: pkg/backend/openshiftcluster/delete.go [hunk 4]
```diff
@@ -24,41 +24,41 @@ func (m *Manager) Delete(ctx context.Context) error {
...
-		if err != nil {
-			return err
-		}
+		for _, subnet := range *nsg.SecurityGroupPropertiesFormat.Subnets {
+			// Note: subnet only has value in the ID field,
+			// so we have to make another API request to get full subnet struct
+			// TODO: there is probably an undesirable race condition here - check if etags can help.
+			s, err := m.subnet.Get(ctx, *subnet.ID)
+			if err != nil {
+				return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidLinkedVNet, "", "Failed to get subnet '%s'. Original error: %s.", *subnet.ID, err.Error())
```
Comment (jim-minter) [inline pkg/backend/openshiftcluster/delete.go:46]: please can you follow the pattern in deployARMTemplate() to return the original error like this:
```
		b, _ := json.Marshal(err)

		return &api.CloudError{
			StatusCode: http.StatusBadRequest,
			CloudErrorBody: &api.CloudErrorBody{
				Code:    xxx,
				Message: "Xxx.",
				Details: []api.CloudErrorBody{
					{
						Message: string(b),
					},
				},
			},
		}
```

File: pkg/backend/openshiftcluster/delete.go [hunk 5]
```diff
@@ -24,41 +26,67 @@ func (m *Manager) Delete(ctx context.Context) error {
...
+			// TODO: there is probably an undesirable race condition here - check if etags can help.
+			s, err := m.subnet.Get(ctx, *subnet.ID)
+			if err != nil {
+				b, _ := json.Marshal(err)
+
+				return &api.CloudError{
+					StatusCode: http.StatusBadRequest,
+					CloudErrorBody: &api.CloudErrorBody{
+						Code:    api.CloudErrorCodeInvalidLinkedVNet,
+						Message: fmt.Sprintf("Failed to get subnet '%s'", *subnet.ID),
```
Comment (jim-minter) [inline pkg/backend/openshiftcluster/delete.go:54]:
```suggestion
						Message: fmt.Sprintf("Failed to get subnet '%s'.", *subnet.ID),
```

File: pkg/backend/openshiftcluster/delete.go [hunk 6]
```diff
@@ -24,41 +26,67 @@ func (m *Manager) Delete(ctx context.Context) error {
...
+			m.log.Printf("removing network security group from subnet %s", *s.ID)
+			err = m.subnet.CreateOrUpdate(ctx, *s.ID, s)
+			if err != nil {
+				b, _ := json.Marshal(err)
+
+				return &api.CloudError{
+					StatusCode: http.StatusBadRequest,
+					CloudErrorBody: &api.CloudErrorBody{
+						Code:    api.CloudErrorCodeInvalidLinkedVNet,
+						Message: fmt.Sprintf("Failed to update subnet '%s'", *subnet.ID),
```
Comment (jim-minter) [inline pkg/backend/openshiftcluster/delete.go:81]:
```suggestion
						Message: fmt.Sprintf("Failed to update subnet '%s'.", *subnet.ID),
```

**Pull request merged**